### PR TITLE
Dev implements provider side of fiveg nrf interface

### PR DESCRIPTION
**Pull request by saltiyazan**
# Description

Implements the provider side of the fiveg_nrf interface

# Checklist:

- [x] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that validate the behaviour of the software
- [x] I validated that new and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [ ] I have bumped the version of the library

### CODE REVIEW

File: tests/unit/test_charm.py [hunk 1]
```diff
@@ -143,3 +143,17 @@ def test_given_container_not_ready_when_database_relation_joins_then_status_is_w
         self.assertEqual(
             self.harness.model.unit.status, WaitingStatus("Waiting for container to be ready")
         )
+
+    def test_given_nrf_urk_when_fiveg_nrf_relation_joined_then_nrf_url_is_in_relation_databag(  # noqa: E501
```
[the remark this answers is not in the document]
Comment (saltiyazan) [reply tests/unit/test_charm.py]: ✅

File: src/charm.py [hunk 2]
```diff
@@ -116,6 +109,31 @@ def _configure_pebble_layer(
         self._container.replan()
         self.unit.status = ActiveStatus()
 
+    def _on_fiveg_nrf_relation_joined(self, event: RelationJoinedEvent) -> None:
+        """Handle fiveg-nrf relation joined event.
+
+        Args:
+            event: RelationJoinedEvent
+        """
+        if self.unit.is_leader():
```
Comment (gruyaume) [inline src/charm.py:118]: Let's follow the pattern we use everywhere else.

```pyhton
if not self.unit.is_leader():
    return
do_what_you_must()
```

File: src/charm.py [hunk 3]
```diff
@@ -116,6 +109,31 @@ def _configure_pebble_layer(
...
+            self.nrf_provider.set_nrf_information(
```
Comment (gruyaume) [inline src/charm.py:119]: I'm not 100% sure this is necessary but I used to check that the nrf service was running before providing it to other NF's. I'm not sure what will happen if we send it to apps and the url/port aren't accessible yet.

Ref [here](https://github.com/gruyaume/nrf-operator/blob/main/src/charm.py#L100)

Comment (saltiyazan) [reply src/charm.py]: I added it, I guess there is no harm in checking anyway :)

File: src/charm.py (hunk 2 above)
Comment (saltiyazan) [reply src/charm.py]: ✅

File: src/charm.py [hunk 4]
```diff
@@ -116,6 +109,31 @@ def _configure_pebble_layer(
...
+                url=NRF_URL,
+                relation_id=event.relation.id,
+            )
+
+    @property
+    def _config_file_is_written(self) -> bool:
```
[the remark this answers is not in the document]
Comment (saltiyazan) [reply src/charm.py:127]: I will address this issue, and some other issues in this [new PR.](https://github.com/canonical/sdcore-nrf-operator/pull/13) 
Bringing some improvements based on the comments on the rest of the sdcore charms.
This is still a draft pr, so don't review it yet. I'm just sharing it to make sure this comment is not lost and we can progress with the current PR idependently.

File: src/charm.py [hunk 5]
```diff
@@ -116,6 +109,31 @@ def _configure_pebble_layer(
...
+        """Returns whether the config file was written to the workload container.
+
+        Returns:
+            bool: Whether the config file was written.
+        """
+        if not self._container.exists(f"{BASE_CONFIG_PATH}/{CONFIG_FILE_NAME}"):
+            logger.info(f"Config file is not written: {CONFIG_FILE_NAME}")
```
[the remark this answers is not in the document]
Comment (saltiyazan) [reply src/charm.py:134]: I will address this issue, and some other issues in this [new PR.](https://github.com/canonical/sdcore-nrf-operator/pull/13) 
Bringing some improvements based on the comments on the rest of the sdcore charms.
This is still a draft pr, so don't review it yet. I'm just sharing it to make sure this comment is not lost and we can progress with the current PR idependently.

File: src/charm.py [hunk 6]
```diff
@@ -116,6 +109,36 @@ def _configure_pebble_layer(
...
+            return False
+        logger.info("Config file is written")
```
Comment (Gmerold) [inline src/charm.py:139]: I think there's no need to log here either. If this function returns `False`, charm will go to Waiting status which is enough I guess.

Comment (saltiyazan) [reply src/charm.py:139]: As above, in the [opened draft pr](https://github.com/canonical/sdcore-nrf-operator/pull/13) I will be completely removing this function :)

File: README.md [hunk 7]
```diff
@@ -18,6 +18,8 @@ Charmed Operator for the SD-CORE Network Repository Function (NRF).
 
 ```bash
 juju deploy sdcore-nrf --trust --channel=edge
+juju deploy mongodb-k8s --trust --channel=5/edge
```
[the remark this answers is not in the document]
Comment (saltiyazan) [reply README.md:21]: If I'm not mistaken the `stable` channel still deploys the legacy charm :)

File: src/charm.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (saltiyazan) [reply src/charm.py:130]: Could you please elaborate a bit more why shouldn't it be a property? I don't see a clear reason :)

File: src/charm.py [hunk 8]
```diff
@@ -173,6 +196,17 @@ def _environment_variables(self) -> dict:
             "MANAGED_BY_CONFIG_POD": "true",
         }
 
+    @property
+    def _nrf_service_is_running(self) -> bool:
+        """Returns whether the NRF service is running."""
```
Comment (sanchezfdezjavier) [inline src/charm.py:201]: Missing return type.

File: tests/unit/test_charm.py [hunk 9]
```diff
@@ -143,3 +143,42 @@ def test_given_container_not_ready_when_database_relation_joins_then_status_is_w
...
+        self.harness.set_leader(is_leader=True)
+        relation_id = self.harness.add_relation(
+            relation_name="fiveg-nrf",
+            remote_app="nrf-requirer",
+        )
+        self.harness.add_relation_unit(relation_id=relation_id, remote_unit_name="nrf-requirer/0")
+        relation_data = self.harness.get_relation_data(
+            relation_id=relation_id, app_or_unit=self.harness.charm.app.name
+        )
+        print(relation_data)
```
Comment (sanchezfdezjavier) [inline tests/unit/test_charm.py:160]:
```suggestion
```

File: tests/unit/test_charm.py [hunk 10]
```diff
@@ -143,3 +143,42 @@ def test_given_container_not_ready_when_database_relation_joins_then_status_is_w
...
+
+        relation_id = self.harness.add_relation(
+            relation_name="fiveg-nrf",
+            remote_app="nrf-requirer",
+        )
+        self.harness.add_relation_unit(relation_id=relation_id, remote_unit_name="nrf-requirer/0")
+        relation_data = self.harness.get_relation_data(
+            relation_id=relation_id, app_or_unit=self.harness.charm.app.name
+        )
+        print(relation_data)
```
Comment (sanchezfdezjavier) [inline tests/unit/test_charm.py:183]:
```suggestion
```

File: metadata.yaml [hunk 11]
```diff
@@ -25,5 +25,9 @@ requires:
   database:
     interface: mongodb_client
 
+provides:  
+  fiveg-nrf:
```
Comment (sanchezfdezjavier) [inline metadata.yaml:29]:
```suggestion
  fiveg_nrf:
```

Comment (saltiyazan) [reply metadata.yaml:29]: I think it's like that in the documentation of the lib :)
https://github.com/canonical/sdcore-nrf-operator/blob/main/lib/charms/sdcore_nrf/v0/fiveg_nrf.py#L40

File: src/charm.py (hunk 8 above)
Comment (saltiyazan) [reply src/charm.py]: ✅

File: src/charm.py [hunk 12]
```diff
@@ -116,6 +109,36 @@ def _configure_pebble_layer(
...
+
+        Args:
+            event: RelationJoinedEvent
+        """
+        if not self.unit.is_leader():
+            return
+        if not self._nrf_service_is_running:
+            event.defer()
+            return
+        print(self._nrf_service_is_running)
```
[the remark this answers is not in the document]
Comment (saltiyazan) [reply src/charm.py]: oh damn! debug leftovers.
Removed!

File: metadata.yaml (hunk 11 above)
Comment (saltiyazan) [reply metadata.yaml:29]: Also in the dummy charms and the test files of the lib.
Let me know if you think it's still better to change it anyway.

File: src/charm.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (gruyaume) [reply src/charm.py:129]: Whether the config file is written or not is not an attribute of the Charm. And we can't "set" it either.

**Comment by saltiyazan** [comment]
> Blocking until the [charm relation interface PR](https://github.com/canonical/charm-relation-interfaces/pull/73) is approved and merged to avoid back and fourths

It's not the same relation interface :) @gruyaume

File: src/charm.py [hunk 13]
```diff
@@ -173,6 +208,21 @@ def _environment_variables(self) -> dict:
             "MANAGED_BY_CONFIG_POD": "true",
         }
 
+    @property
```
Comment (gruyaume) [inline src/charm.py:211]: Also probably not a property

**Review by sanchezfdezjavier — APPROVED** [review]

File: src/charm.py (hunk 6 above)
Comment (Gmerold) [reply src/charm.py:151]: Nope, that's fine.

**Review by Gmerold — APPROVED** [review]

**Pull request merged**